File: solutions/csharp/trafficsignalsystem/States/Intersection/NorthSouthGreenState.cs
class NorthSouthGreenState : IIntersectionState
{
    public void Handle(IntersectionController context)
    {
        Console.WriteLine($"\n--- INTERSECTION {context.GetId()}: Cycle Start -> North-South GREEN ---");

        // Turn North and South green, ensure East and West are red
        context.GetLight(Direction.NORTH).StartGreen();
        context.GetLight(Direction.SOUTH).StartGreen();
        context.GetLight(Direction.EAST).SetColor(LightColor.RED);
        context.GetLight(Direction.WEST).SetColor(LightColor.RED);

        // Wait for green light duration
        Thread.Sleep((int)context.GetGreenDuration());

        // Transition North and South to Yellow
        context.GetLight(Direction.NORTH).Transition();
        context.GetLight(Direction.SOUTH).Transition();

        // Wait for yellow light duration
        Thread.Sleep((int)context.GetYellowDuration());

        // Transition North and South to Red
        context.GetLight(Direction.NORTH).Transition();
        context.GetLight(Direction.SOUTH).Transition();

        // Change the intersection's state to let East-West go
        context.SetState(new EastWestGreenState());
    }
}

File: solutions/java/src/trafficsignalcontrolsystem/states/intersection/NorthSouthGreenState.java
package trafficsignalcontrolsystem.states.intersection;

import trafficsignalcontrolsystem.IntersectionController;
import trafficsignalcontrolsystem.enums.Direction;
import trafficsignalcontrolsystem.enums.LightColor;

public class NorthSouthGreenState implements IntersectionState {
    @Override
    public void handle(IntersectionController context) throws InterruptedException {
        System.out.printf("\n--- INTERSECTION %d: Cycle Start -> North-South GREEN ---\n", context.getId());

        // Turn North and South green, ensure East and West are red
        context.getLight(Direction.NORTH).startGreen();
        context.getLight(Direction.SOUTH).startGreen();
        context.getLight(Direction.EAST).setColor(LightColor.RED);
        context.getLight(Direction.WEST).setColor(LightColor.RED);

        // Wait for green light duration
        Thread.sleep(context.getGreenDuration());

        // Transition North and South to Yellow
        context.getLight(Direction.NORTH).transition();
        context.getLight(Direction.SOUTH).transition();

        // Wait for yellow light duration
        Thread.sleep(context.getYellowDuration());

        // Transition North and South to Red
        context.getLight(Direction.NORTH).transition();
        context.getLight(Direction.SOUTH).transition();

        // Change the intersection's state to let East-West go
        context.setState(new EastWestGreenState());
    }
}
